File: cookbook/python-a2a/nodes.py
import yaml
from caskada import Context, node
from utils import call_llm, search_web


@node
def decide(context: Context) -> None:
    print("🤔 Agent deciding what to do next...")
    response = call_llm(
        f"""
You are a research assistant that can search the web.
Question: {context.state["question"]}
Previous research: {context.state.get("research", "No previous search")}

## NEXT ACTION
Choose `search` or `answer` and return exactly one of these YAML code blocks:
```yaml
action: search
search_query: what to search for
```

```yaml
action: answer
answer: final answer when action is answer
```
"""
    )
    if "```yaml" not in response:
        raise ValueError("decision must contain a YAML block")
    decision = yaml.safe_load(response.split("```yaml", 1)[1].split("```", 1)[0])
    if not isinstance(decision, dict) or decision.get("action") not in {
        "search",
        "answer",
    }:
        raise ValueError("decision action must be search or answer")

    if decision["action"] == "search":
        search_query = decision.get("search_query")
        if not isinstance(search_query, str) or not search_query.strip():
            raise ValueError("search decision needs a non-empty search_query")
        context.state["search_query"] = search_query
        print(f"🔍 Agent decided to search for: {search_query}")
    else:
        answer = decision.get("answer")
        if not isinstance(answer, str) or not answer.strip():
            raise ValueError("answer decision needs a non-empty answer")
        context.state["research"] = answer
        print("💡 Agent decided to answer the question")
    context.emit(decision["action"])


@node
def search(context: Context) -> None:
    query = context.state["search_query"]
    print(f"🌐 Searching the web for: {query}")
    results = search_web(query)
    context.state["research"] = (
        context.state.get("research", "") + f"\n\nSEARCH: {query}\nRESULTS: {results}"
    )
    print("📚 Found information, analyzing results...")
    context.emit("decide")


@node
def answer(context: Context) -> None:
    print("✍️ Crafting final answer...")
    context.state["answer"] = call_llm(
        f"""
Answer this question using the research below.
Question: {context.state["question"]}
Research: {context.state.get("research", "")}
"""
    )
    print("✅ Answer generated successfully")

File: cookbook/typescript-agent/nodes.ts
import { node } from 'sley'
import { parse } from 'yaml'
import { callLLM, webSearch } from './utils'

import type { AgentState, Decision } from './types'

function parseDecision(response: string): Decision {
  const yaml = response.split('```yaml', 2)[1]?.split('```', 1)[0]
  if (yaml === undefined) throw new Error('LLM response must contain a YAML block')

  const value: unknown = parse(yaml)
  if (typeof value !== 'object' || value === null) {
    throw new Error('LLM decision must be a YAML object')
  }

  const decision = value as Record<string, unknown>
  if ((decision.action !== 'search' && decision.action !== 'answer') || typeof decision.reason !== 'string') {
    throw new Error('LLM decision needs a valid action and reason')
  }

  if (decision.action === 'search') {
    if (typeof decision.searchQuery !== 'string' || decision.searchQuery.trim() === '') {
      throw new Error('A search decision needs a non-empty searchQuery')
    }
    return {
      action: 'search',
      reason: decision.reason,
      searchQuery: decision.searchQuery,
    }
  }

  return { action: 'answer', reason: decision.reason }
}

export const decide = node<AgentState>(async (context) => {
  console.log('Deciding whether to search or answer the question...')
  const response = await callLLM(`
### Context
Current date: ${new Date().toISOString()}
Question: ${context.state.question}
Research: ${context.state.research ?? 'No previous search.'}

### Next Action
Choose whether to search for more information or answer now.

Return exactly one YAML code block in one of these forms:

\`\`\`yaml
action: search
reason: Why another search is needed
searchQuery: A concise query written for a search engine
\`\`\`

\`\`\`yaml
action: answer
reason: Why the available research is sufficient
\`\`\`

Use only the keys shown for the selected action. Do not write text outside the block.
`)
  const decision = parseDecision(response)
  console.log(`Reason: ${decision.reason}`)

  if (decision.action === 'search') {
    context.state.searchQuery = decision.searchQuery
    console.log(`Searching for: ${decision.searchQuery}`)
  } else {
    console.log('Research is sufficient; preparing the final answer.')
  }

  context.emit(decision.action)
})

export const search = node<AgentState>(async (context) => {
  const query = context.state.searchQuery
  if (query === undefined) throw new Error('searchQuery is missing')

  console.log('Calling web search tool.')
  const results = await webSearch(query)
  context.state.research = `${context.state.research ?? ''}\n\nSearch: ${query}\nResults: ${JSON.stringify(results)}`
  context.emit('decide')
})

export const answer = node<AgentState>(async (context) => {
  const response = await callLLM(`
Answer this question using the research below.

Question: ${context.state.question}
Research: ${context.state.research ?? 'No previous search.'}
`)
  context.state.answer = response
  console.log(`Final Answer: ${response}`)
})
